File: android/src/main/kotlin/com/crdzbird/wearer_link/BackgroundDispatcher.kt
package com.crdzbird.wearer_link

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.FlutterInjector
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.dart.DartExecutor
import io.flutter.view.FlutterCallbackInformation

/**
 * Headless Dart delivery for events that arrive while no UI engine exists.
 *
 * Flow: [WearerLinkListenerService] persists the event (crash-safe), then
 * hands it here. A background [FlutterEngine] is started (once per process)
 * on the plugin's dispatcher entrypoint; after the isolate's
 * `backgroundReady` handshake each event is pushed over
 * [WearerLinkBackgroundFlutterApi] and removed from the persistent queue
 * only when the Dart handler's future completes — an unhandled handler
 * error leaves it queued for the next foreground launch (at-least-once).
 */
internal object BackgroundDispatcher : WearerLinkBackgroundHostApi {

  private const val PREFS = "wearer_link_background"
  private const val KEY_DISPATCHER = "dispatcher_handle"
  private const val KEY_USER = "user_handle"

  private val mainHandler = Handler(Looper.getMainLooper())

  // All mutable state below is confined to the main thread.
  /**
   * True while the headless engine is being constructed. FlutterEngine
   * auto-registers all plugins (including WearerLinkPlugin) during
   * construction; this flag stops that instance from claiming the live
   * dispatcher slot meant for UI engines. Main-thread confined.
   */
  var creatingBackgroundEngine = false
    private set

  private var engine: FlutterEngine? = null
  private var api: WearerLinkBackgroundFlutterApi? = null
  private var ready = false
  private val awaitingReady = ArrayDeque<WearerEventDto>()
  private var appContext: Context? = null

  fun register(context: Context, dispatcherHandle: Long, userHandle: Long) {
    prefs(context).edit()
      .putLong(KEY_DISPATCHER, dispatcherHandle)
      .putLong(KEY_USER, userHandle)
      .apply()
  }

  fun clear(context: Context) {
    prefs(context).edit().clear().apply()
  }

  fun isRegistered(context: Context): Boolean =
    prefs(context).getLong(KEY_DISPATCHER, 0L) != 0L

  /**
   * Deliver [dto] to the background isolate, starting it if needed.
   * The event must already be persisted; it is removed on ack.
   */
  fun deliver(context: Context, dto: WearerEventDto) {
    val app = context.applicationContext
    mainHandler.post {
      appContext = app
      if (ready) {
        push(app, dto)
      } else {
        awaitingReady.add(dto)
        startEngineIfNeeded(app)
      }
    }
  }

  // -- WearerLinkBackgroundHostApi (called from the background isolate) -----

  override fun backgroundReady(): Long {
    val context = appContext
      ?: throw FlutterError("unknown", "Background engine without context.", null)
    ready = true
    // Flush on the NEXT main-loop turn: platform messages are delivered in
    // order, so events pushed inside this handler would reach the isolate
    // before the backgroundReady reply it needs to resolve the user handler.
    mainHandler.post {
      while (awaitingReady.isNotEmpty()) {
        push(context, awaitingReady.removeFirst())
      }
    }
    return prefs(context).getLong(KEY_USER, 0L)
  }

  // -- internals ------------------------------------------------------------

  private fun startEngineIfNeeded(context: Context) {
    if (engine != null) return
    val handle = prefs(context).getLong(KEY_DISPATCHER, 0L)
    if (handle == 0L) return
    // The loader must be fully initialized BEFORE the callback lookup:
    // lookupCallbackInformation is a native call into libflutter.so.
    val loader = FlutterInjector.instance().flutterLoader()
    if (!loader.initialized()) {
      loader.startInitialization(context)
    }
    loader.ensureInitializationComplete(context, null)
    val callback = FlutterCallbackInformation.lookupCallbackInformation(handle)
      ?: return // stale handle from a previous app version; queue keeps the events
    creatingBackgroundEngine = true
    val backgroundEngine = try {
      FlutterEngine(context)
    } finally {
      creatingBackgroundEngine = false
    }
    WearerLinkBackgroundHostApi.setUp(
      backgroundEngine.dartExecutor.binaryMessenger,
      this,
    )
    api = WearerLinkBackgroundFlutterApi(backgroundEngine.dartExecutor.binaryMessenger)
    backgroundEngine.dartExecutor.executeDartCallback(
      DartExecutor.DartCallback(context.assets, loader.findAppBundlePath(), callback),
    )
    engine = backgroundEngine
  }

  private fun push(context: Context, dto: WearerEventDto) {
    val activeApi = api ?: return
    activeApi.onBackgroundEvent(dto) { result ->
      if (result.isSuccess) {
        PendingEventStore(context).remove(dto.id)
      }
      // Failure: the handler threw or the isolate died — the event stays in
      // the persistent queue and replays on the next foreground launch.
    }
  }

  private fun prefs(context: Context) =
    context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
}

File: ios/wearer_link/Sources/wearer_link/BackgroundDispatcher.swift
import Flutter
import Foundation

/// Headless Dart delivery for events that arrive while no UI engine exists
/// (watch-triggered background launch). Mirrors the Android
/// BackgroundDispatcher: events are persisted first, pushed into a headless
/// FlutterEngine running the plugin's dispatcher entrypoint, and removed
/// from the queue only when the Dart handler's future completes.
final class BackgroundDispatcher: NSObject {
  static let shared = BackgroundDispatcher()

  private static let keyDispatcher = "wearer_link_bg_dispatcher_handle"
  private static let keyUser = "wearer_link_bg_user_handle"

  private let defaults = UserDefaults.standard

  // Main-thread confined.
  private var engine: FlutterEngine?
  private var api: WearerLinkBackgroundFlutterApi?
  private var hostApi: WearerLinkPlugin?
  private var ready = false
  private var awaitingReady: [WearerEventDto] = []

  private override init() { super.init() }

  func register(dispatcherHandle: Int64, userHandle: Int64) {
    defaults.set(dispatcherHandle, forKey: Self.keyDispatcher)
    defaults.set(userHandle, forKey: Self.keyUser)
  }

  func clear() {
    defaults.removeObject(forKey: Self.keyDispatcher)
    defaults.removeObject(forKey: Self.keyUser)
  }

  var isRegistered: Bool {
    (defaults.object(forKey: Self.keyDispatcher) as? Int64 ?? 0) != 0
  }

  /// Deliver an already-persisted event to the background isolate,
  /// starting the headless engine if needed. Call on the main thread.
  func deliver(_ event: WearerEventDto) {
    if ready {
      push(event)
    } else {
      awaitingReady.append(event)
      startEngineIfNeeded()
    }
  }

  // MARK: - internals

  private func startEngineIfNeeded() {
    guard engine == nil else { return }
    let handle = defaults.object(forKey: Self.keyDispatcher) as? Int64 ?? 0
    guard handle != 0,
      let info = FlutterCallbackCache.lookupCallbackInformation(handle)
    else { return } // stale handle; the persistent queue keeps the events
    let backgroundEngine = FlutterEngine(
      name: "wearer_link_background", project: nil, allowHeadlessExecution: true)
    guard backgroundEngine.run(
      withEntrypoint: info.callbackName, libraryURI: info.callbackLibraryPath)
    else { return }
    WearerLinkBackgroundHostApiSetup.setUp(
      binaryMessenger: backgroundEngine.binaryMessenger, api: self)
    // Expose the regular host API too, so the background handler can
    // send/sync back to the watch. Only the API — live event dispatch stays
    // with the UI engine; this isolate is fed via the background channel.
    let host = WearerLinkPlugin()
    WearerLinkHostApiSetup.setUp(
      binaryMessenger: backgroundEngine.binaryMessenger, api: host)
    hostApi = host
    api = WearerLinkBackgroundFlutterApi(
      binaryMessenger: backgroundEngine.binaryMessenger)
    engine = backgroundEngine
  }

  private func push(_ event: WearerEventDto) {
    api?.onBackgroundEvent(event: event) { result in
      if case .success = result {
        PendingEventStore.shared.remove(id: event.id)
        StatsStore.shared.increment(StatsStore.keyBackground)
      }
      // Failure: handler threw — the event stays queued for the next launch.
    }
  }
}

// MARK: - WearerLinkBackgroundHostApi

extension BackgroundDispatcher: WearerLinkBackgroundHostApi {
  func backgroundReady() throws -> Int64 {
    ready = true
    // Flush on the NEXT main-loop turn: events pushed inside this handler
    // would reach the isolate before the backgroundReady reply it needs to
    // resolve the user handler (platform messages are delivered in order).
    DispatchQueue.main.async {
      let queued = self.awaitingReady
      self.awaitingReady = []
      queued.forEach(self.push)
    }
    return defaults.object(forKey: Self.keyUser) as? Int64 ?? 0
  }
}
